src/algorithms/policy_gradient/A2CGAE.jl

```diff
@@ -0,0 +1,84 @@
+export A2CGAELearner
+
+using Flux
+
+"""
+    A2CGAELearner(;kwargs...)
+
+# Keyword arguments
+
+- `approximator`, an [`ActorCritic`](@ref) based [`NeuralNetworkApproximator`](@ref)
+- `γ::Float32`, reward discount rate.
+- 'λ::Float32', lambda for GAE-Lamda
+- `actor_loss_weight::Float32`
+- `critic_loss_weight::Float32`
+- `entropy_loss_weight::Float32`
+"""
+Base.@kwdef struct A2CGAELearner{A} <: AbstractLearner
+    approximator::A
+    γ::Float32
+    λ::Float32
+    actor_loss_weight::Float32
+    critic_loss_weight::Float32
+    entropy_loss_weight::Float32
+end
+
+(learner::A2CGAELearner)(obs::BatchObs) =
+    learner.approximator(
+        send_to_device(device(learner.approximator), get_state(obs)),
+        Val(:Q),
+    ) |> send_to_host
+
+function RLBase.update!(learner::A2CGAELearner, experience)
+    AC = learner.approximator
+    γ = learner.γ
+    λ = learner.λ
+    w₁ = learner.actor_loss_weight
+    w₂ = learner.critic_loss_weight
+    w₃ = learner.entropy_loss_weight
+    states, actions, rewards, terminals, next_state = experience
+    states = send_to_device(device(AC), states)
+    next_state = send_to_device(device(AC), next_state)
+
+    states_flattened = flatten_batch(states) # (state_size..., n_thread * update_step)
+    actions = flatten_batch(actions)
+    actions = CartesianIndex.(actions, 1:length(actions))
+
+    next_state_values = AC(next_state, Val(:V))
+    gains = discount_rewards(
+        rewards,
+        γ*λ;  # discounting the advantage estimation by λγ
+        dims = 2,
+        init = send_to_host(next_state_values),
+        terminal = terminals,
+    )
```

Comment:
The `discount_rewards` function doesn't implement the GAE here, though they do share some common codes. I'd suggest you take a look at the source code of [`discount_rewards`](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/master/src/utils/base.jl#L194-L206) and figure out how to write a GAE function.

Reply:
According to the **Eq 16** in [the original paper](https://arxiv.org/pdf/1506.02438.pdf), a GAE function may need to accept the following arguments:

- `rewards`
- `values`
- `λ`
- `γ`
- `terminal`

And to support 2-D array, we'd better add a `dims` keyword argument here.

You can start from the 1-D version first.

Reply:
Ref: https://github.com/FortsAndMills/Learning-Reinforcement-Learning/blob/master/LRL/GAE.py#L26-L29